### ISSUE TITLE
Flyback protection won't work as advertised.

### ISSUE THREAD

**Issue by dead-lock-ed** [condottab/Roborock-CPAP#3]
Forward based Schottky diode won't protect from inductive spike from switching an inductor. Instead it'll just hold the charge until MOSFET or diode dies. This is shown by the inductive spike from simulated switching in Falstad Circuit Simulator. In reality the spike won't be as high due to capacitances and rising time of the signals, but still there's a possibility for the spike.
![Flyback comparison](https://github.com/condottab/Roborock-CPAP/assets/53382462/35acd5c5-be63-4eaa-bef0-268f4e1bab1d)
Hotend / fan ports on mainboards in 3D printers switch low side (GND is switched instead of positive voltage rail) meaning that this PCB which is referencing protection of PWM pin via Zener diodes to GND won't function properly. And even if they did, ZENER1 has no extra series resistance meaning it'd die from a spike too big.
Another valid concern is expected PWM voltage of the ESC as PWM port from mainboard will provide 3,3 V, due to forward drop over the Schottky2 it'll be a bit below 3 V, then it's going through a pull-down + forward resistor configuration which is made into a voltage divider, meaning that ESC will see 2,7 V, while no signal voltage range is provided here in documentation.

**Comment by TheRealPunkt** [comment]
you flyback on the motor + is wrong you use a dc motor the nidec ares bldc with internal esc so there is no real pwm on the + rail whatcan be happen a small drop pseudo pwm to 22v better is use a oszi an meassure befor the diode to check it

simulation are not  allday correct when you wanna simulate it correct you need to build a correct bldc motor up

**Comment by dead-lock-ed** [comment]
> you flyback on the motor + is wrong

Where? I'm showing comparison of the solution proposed by this repo (on the left) to the most common flyback protection method (on the right)
> you use a dc motor the nidec ares bldc with internal esc

While true, for transient analysis they're basically the same, they both have resistance and inductance, torque and electromagnetic constants, in some cases you can make BDC out of BLDC and the other way around as they just have different commutation method 
> so there is no real pwm on the + rail 

There's no PWM on the supply, but circuit of this repo switches the power to entire ESC low side as otherwise the ESC goes into error state according to the Roborock CPAP PCB maintainers, this section of analysis was exactly about possible transient spike from this event and if the circuit can protect it if it goes past ESC protection which was never determined 
> whatcan be happen a small drop pseudo pwm to 22v better is use a oszi an meassure befor the diode to check it

No need to, if you see transient response before the diode in proposed circuit it means that it failed, also proposed Schottky diode wouldn't give 2 V forward voltage in predicted operating conditions 
> simulation are not allday correct when you wanna simulate it correct you need to build a correct bldc motor up

There's no need to if you build a current circuit instead. ESC in this scenario is a black box, no one even checked what the ESC model is nor attempted to RE the digital section of the mainboard this ESC connects to.

**Comment by dbuezas** [comment]
I just tested the fan with the oscilloscope and I don't see any back EMF on the PWM pin whatsoever.

**Comment by dead-lock-ed** [comment]
It's an intermittent issue, these spikes with this driver shouldn't be common but they can occur if the fan is controlled this way, and if they happen, the board has no real way of providing protection. 
If there's no spike, why use the board at all?
If there's a spike, this board won't protect from it.

**Comment by dbuezas** [comment]
What makes me doubt everything is that the fan's purpose is to be driven by the mcu of the robot vacuum. Why would they design the pcb of the fan in such a way that it requires so much care to drive safely if it already has a driver

**Comment by dead-lock-ed** [comment]
Because we aren't using the PCB as intended. The risk comes from us switching the low side without any concern for the inner workings of the ESC as no one managed to find the datasheet of the IC used.

**Comment by dbuezas** [comment]
Oh, I see. So even though the pwm pin will set the speed, since gnd will be left floating when the fan is off there's the risk of overvolting the pwm pin. Gotcha.

I'll use it with Marlin so I'll have to be creative

**Comment by dbuezas** [comment]
I've been playing a bit with the fan and both the tacho and pwm pins go to 21 v when gnd is left floating.
Regardless of MCU pin protection, having some current going from vcc to the pwm pin internally is likely bad for the fan driver.

I'm using a 2€ relay board to switch the high side instead with a gpio and it seems to be working very well.
Unfortunately Marlin doesn't have a configuration for a fan enable pin, so I'll either make a PR or make a small circuit to "derive" the enable pin from the pwm one.

**Comment by dead-lock-ed** [comment]
Thank you for looking more into it.
Indeed, switching high side with a relay or a P-MOSFET is a more optimal solution IMO. I'm fairly certain that some Marlin folks would gladly include fan enable, you could also try to make macros called at the start and end of the print to turn on and off the relay and instead of turning the fan off completely during the print, just set the speed that it'll be barely on or in case of some ESCs there's a set of setpoints where it won't even try to turn the fan, it'll just be set to idle.
Good luck!

**Comment by dbuezas** [comment]
Good Idea with the pmos high side switch, I'm afraid it heats up too much, but I'll give it a try.
I'm thinking of using a bjt to switch the gate of the pmos (a rather typical approach). I'll report back.

I'm also testing the pwm frequency, 20kHz works fine without noise, do you know if there's any risk with high pwm frequencies?

**Comment by dbuezas** [comment]
![Screenshot_20240423-184306~2](https://github.com/condottab/Roborock-CPAP/assets/777196/02232321-7fbf-4ba8-93fb-ab8a0bf4c655)

It works in the simulator. The fan is represented in blue, the enable signal is generated from the pwm by rectifying and buffering it. The BJT pulls Vgs to -10 to turn on the high side switch.

I'll give it a try on the breadboard soon. 

**Update: works like a charm, but add a 10k resistor between GND and PWM to avoid it turning on on boot**

**Comment by dead-lock-ed** [comment]
Out of curiosity, why only -10 V? Can the P-MOSFET handle more? Other than that, it's looking solid.
As for the higher frequencies, sadly I don't know anything, as noted previously, there seems to be no documentation about used IC at all. If you want to try your own luck, it's labeled BD6921.

**Comment by ripinteer** [comment]
> ![Screenshot_20240423-184306~2](https://private-user-images.githubusercontent.com/777196/324920884-02232321-7fbf-4ba8-93fb-ab8a0bf4c655.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM4OTI5MTAsIm5iZiI6MTcxMzg5MjYxMCwicGF0aCI6Ii83NzcxOTYvMzI0OTIwODg0LTAyMjMyMzIxLTdmYmYtNGJhOC05M2ZiLWFiOGEwYmY0YzY1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyM1QxNzE2NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOTZjYzNkM2QwN2UxYWM2NzA2YjRkOTgwODRjNTZmYmNjODA5YTg0MGQ5NGE2YWZmNTYxYzk3NjcwZTE2OTUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.3imtw9_9yORCXdGuOf8QtgGr7evMH5ISZAb6MfpTDcw)
> 
> It works in the simulator. The fan is represented in blue, the enable signal is generated from the pwm by rectifying and buffering it. The BJT pulls Vgs to -10 to turn on the high side switch.
> 
> I'll give it a try on the breadboard soon.

I have successfully implemented this idea in one of my pcbs for similar purpose. Although, N-channel MOSFET was utilised. 
I never thought about P-channel driving, despite some tricky moments I’ve encountered in testing, related to missing ground. 
As for “spikes” concerns at PWM line - me (and users as well) never experienced these issues so far

**Comment by dbuezas** [comment]
> Out of curiosity, why only -10 V? Can the P-MOSFET handle more? 

The datasheet says that the absolute max Vgs is -25V, and all the Rds plots go up to -10v so the reason is just to be on the safe side. By -10 it is already fully saturated anyway.

> it's labeled BD6921.

Thanks for the tip!

**Comment by dead-lock-ed** [comment]
> As for “spikes” concerns at PWM line - me (and users as well) never experienced these issues so far

I think that I know of at least 2 cases where mainboard pin used to PWM died while using this board. 
So we have proof that it doesn't protect. 
And we have a model showing that it can't really protect either way.

**Comment by ripinteer** [comment]
> > As for “spikes” concerns at PWM line - me (and users as well) never experienced these issues so far
> 
> I think that I know of at least 2 cases where mainboard pin used to PWM died while using this board. So we have proof that it doesn't protect. And we have a model showing that it can't really protect either way.

I should’ve pointed out that I am speaking from experience mainly not with this scheme. I initially flopped on this one, and as a result designed my own with TWS diodes on pwm/tach lines, for “just in case”.
Both of diodes them turned up useless 

Now I am working on a quite different approach, instead of mounting pcb in Polulu socket (which I won’t deprecate though), I will different format inspired by Jniebuhr’s one

**Comment by dbuezas** [comment]
The high switch from pwm circuit above works perfectly.
![17138998632833882405097841829741](https://github.com/condottab/Roborock-CPAP/assets/777196/4647b6fd-4ec5-48ba-97e4-3177fc9a84e2)

I used an FQP27P06 pmos and a 2n2222. Any npn bjt would work, but the pmos needs to be able to handle the 24v (this one can do 60v Vds and -24 Vgs, with 0.075ohms)

The power switches on from even very low duty cycle (1%) at high frequency (55kHz). It also cuts vcc fine in less than half a second when pwm is turned off.
The bare mosfet gets a bit warm above 30W of power (>70%), but that's already way beyond what the vacuum cleaner uses and ludicrous cooling power.

I also see zero spikes in any pin (pwm or tacho). Neither negative nor positive voltages on my oscilloscope.

Only takes 1 normal "unprotected" pwm input direct from the mcu and needs no extra enable pin.

I like it more than the relay, I'll put it in a perf board in the coming days and move it to the printer.

So, only 1 "unprotected" pwm pin used from the mcu.

BTW @dead-lock-ed , the driver turns off whenever the fan stops moving for a small period of time, so keeping PWM very low instead of turning off doesn't avoid the error state: the power needs to be cut off (here by the high side switch) for it to recover. Obviously a good safety measure.

Also in the datasheet of a similar (FAN6921 just different manufacturer?) chip, they suggest a typical frequency of 55kHz or anything above 20kHz to avoid noise.

**Comment by dbuezas** [comment]
![Screenshot_20240423-215815](https://github.com/condottab/Roborock-CPAP/assets/777196/0430d8b6-b8a2-4fba-84f8-d8814b80a5e4)

I didn't dig deeper in the datasheet, but I think this is about stall detection, and I interpret it as "turn it off and on again", exactly as we do.

**Comment by dbuezas** [comment]
@ripinteer 

> As for “spikes” concerns at PWM line - me (and users as well) never experienced these issues so far

I disconnected gnd and saw ~20v in the pwm line. The mcu or some zener diode protection may be able to sink it fine, but it is probably risky. Plus pulling current from the pwm input of the fan's PCB sounds icky to me. Maybe it's fine for a while, maybe forever, but I'd personally just switch vcc instead and avoid undefined behavior.

**Comment by dbuezas** [comment]
Also I just noticed the protection circuit in this repo has the zeners connected to the switched "gnd" which is floating when off. That's not going to do much to protect the mcu

**Comment by dead-lock-ed** [comment]
@dbuezas FAN6921 doesn't seem similar at all, it's for PFC and power conditioning.
Glad to hear that the P-MOSFET circuit is working.
Yeah, I mentioned this Zener reference to fake GND issue in the original comment as well, overall the circuit in this repo has to be redesigned from the ground up, so far I've been gathering the characteristics of the fan from a few sources.

**Comment by dbuezas** [comment]
![IMG-20240425-WA0006](https://github.com/condottab/Roborock-CPAP/assets/777196/b1e06bab-f10f-4dcd-9b2c-914b7689df0b)

I think my Chinese mosfets may be counterfeit. Anyway, just increased Vgs (more negative that is) and added a second one to reduce Rds on.

Probably unnecessary but it now runs cooler than the fan's PCB anyway. I measure ~600mV of loss in the PMOS pair at 20V and 20W. So 1A*.6v= 0.6W of loss = 97% efficiency.

Relay is probably better but bigger and it clicks

**Comment by dbuezas** [comment]
Also the tacho is fun, I get 120,000 rpm before my puny power supply turns off. With the high side switch it is safe to read it too (Marlin at least can)

**Comment by dead-lock-ed** [comment]
They claim 70 mΩ Rdson so your result suggesting 1,2 Ω Rdson is indeed quite worrying. Still should be perfectly fine though

**Comment by dbuezas** [comment]
Yeah and that at Vgs = -16v instead of the 10 the datasheet suggests.
Most likely fake transistors. AliExpress for the win! They're not even new 🤣

**Comment by dbuezas** [comment]
I finally started installing it in the printer. The fan duct on an ultimaker was quite the puzzle, particularly one with a direct drive extruder and an ABL probe.
<img width="297" alt="image" src="https://github.com/condottab/Roborock-CPAP/assets/777196/b723270a-c57e-4089-b029-bda6f24572e6">

## The PWM pin NEEDS a 10k pulldown

otherwise the fan goes to 100% on boot. And 100% of this thing at 24v is no joke.

I'll repeat: If anybody wants to use the circuit I posted **add a 10k resistor between PWM and GND**

**Comment by dbuezas** [comment]
Fun fact: I wired and configured the tacho measurement and in Marlin it is useless with this fan. It can only measure up to 7000RPM, which is a tenth of what the fan can do without pushing it.

**Comment by dead-lock-ed** [comment]
@dbuezas do you perhaps know if that's a Marlin limitation or just tacho on the ESC being useless?

**Comment by dbuezas** [comment]
Marlin. I could read it fine with an esp32 running esphome. Marlin probably just polls the pin at a fixed rate

**Comment by condottab** [comment]
Sorry for the delay in responding. To my knowledge no MCUs or MOSFETs have died from the fan with this protection, I know of one board that fried when the user connected and disconnected the signal pin online. AFAIK hundreds of people are currently using this type of board to run their fans and I think I would be getting some negative feedback at this point if it was truly unsafe. If you think the protection is not sufficient - feel free to post your version, I would be happy to discuss your results. As for the signal range, it's 2-7V.

**Comment by dbuezas** [comment]
My guess with the fried board is that the pwm input line of the fan goes to 24v but is fairly high impedance, so if it is kept connected, the mcu will keep it down, but if hot plugged will basically shock it

**Comment by dead-lock-ed** [comment]
> To my knowledge no MCUs or MOSFETs have died from the fan with this protection

I know of 2 cases where the specifically MCU PWM pin got fried during normal usage

> I think I would be getting some negative feedback at this point if it was truly unsafe.

I started my investigations because of one person that got their pin fried and few weeks later I got contacted by another person whose board got fried.

> If you think the protection is not sufficient - feel free to post your version, I would be happy to discuss your results. 

That's what I'm in process of doing, but I have a lot still to learn. I have a lot to consider and I don't want to release anything that would be insufficient, while still being fairly low cost.

**Comment by condottab** [comment]
I find it odd that people are contacting you and not me. My unit has been working fine all these months, not to mention the dozens of people in my close circle of hobby friends. I am willing to leave this question open as people need to know your opinion.

**Comment by dead-lock-ed** [comment]
I got approached the first time just to troubleshoot the mainboard as it suddenly stopped turning the CPAP on.
As I noted previously, this spike shouldn't ever exist if the ESC is operating as it should. But in case it somehow happens, this board isn't built to protect from it.

**Comment by condottab** [comment]
This circuit was created after a back emf killed an MCU. As far as I know, the protection did not cause any damage since.

**Comment by dbuezas** [comment]
A 4v zener between pwm and real gnd should be enough for mcu protection. This isn't good for the fan's pcb though, it will basically see -20v on an input (although without a gnd reference it's all out of spec)

The alternative i choose (the circuit above) is to switch vcc instead and extract the enable signal from the pwm, which reduces the mcu pin count and removes the need of changing marlin's source code. Since the fan is always referenced to real gnd, everything behaves reasonably including the tacho pin.

**Comment by dbuezas** [comment]
I designed and ordered a board for the fan, including a frequency divider to one can read the tacho in marlin.
If anybody interested, lmk!

<img width="700" alt="Image" src="https://github.com/user-attachments/assets/1ffa9311-1695-4e4b-a822-301b393d80bb" />

**Comment by dead-lock-ed** [comment]
Looks great, I'm also digging the looks of curved traces on this PCB. Keep us posted

**Comment by dbuezas** [comment]
Boards arrived!

![Image](https://github.com/user-attachments/assets/5dfe03b7-580f-489e-8243-46593fc3631a)

![Image](https://github.com/user-attachments/assets/ffb1d01c-679c-4f79-84d9-b5c2179d832e)

After some crimping and changing the screws for 3 longer ones (with nuts as standoffs):

![Image](https://github.com/user-attachments/assets/f2ecb04a-1e53-41d6-800e-560d31080bfd)

Tacho divider works as intended

![Image](https://github.com/user-attachments/assets/29bfd9e5-7ec9-480b-a98d-75e733780ff1)

And the tachometer is read correctly by marlin.
![Image](https://github.com/user-attachments/assets/54861528-485b-4df6-adad-518382c0b635)

Everything works as intended, the pmos don't heat at all and there are no voltage spikes anywhere (since i'm switching vcc, not gnd)

I have 4 boards I could sell if anybody interested. 5€ + shipping just to cover costs. I'll upload everything to my github in a couple of weeks.

**Comment by MahlioPaluori** [comment]
If only i have found your repo earlier
Been trying to use the nidec fan on my printer with various solutions, burned one ESC and ended with the next realisation:
- voltage divider with 1k/250 Ohm resistiors on the partcooling fan port of MB to get ~5v PWM for control
- one of the gpios from MB (i have a btt skr pico with rp2040 MCU) to switch the SSR on the high line to power the nidec fan
- separate PSU to power the nidec fan, cause having all outputs on the MB with low-switching circuits was giving me constant +24V between the PWM and any other GND, so the fan wasn\t reacting to PWM and spinning at 100%
- new macro for m107/m107 with 0.1 sec delay between applying pwm and power to the fan, cause it says that it needs PWM signal prior to power

works fine as described
